cni/pkg/nodeagent/informers.go

```diff
@@ -260,45 +284,66 @@ func (s *InformerHandlers) reconcilePod(input any) error {
 			log.Debugf("removing pod from mesh: no longer should be enabled")
 			err := s.dataplane.RemovePodFromMesh(s.ctx, pod, false)
 			log.Debugf("RemovePodFromMesh returned: %v", err)
+			return err
 			// we ignore errors here as we don't want this event to be retried by the queue.
-		} else {
-			// If oldpod != ready && newpod != ready, but the ambient annotation was added,
-			// then assume this event was generated by the CNI plugin labeling the pod on startup,
-			// and skip the event.
-			//
-			// This isn't perfect (someone could manually annotate an unready pod,
-			// then install Istio, then the pod goes ready, and we'd miss capture) - but that
-			// seems vanishingly unlikely
-			wasReady := kube.CheckPodReadyOrComplete(oldPod)
-			isReady := kube.CheckPodReadyOrComplete(newPod)
-			if wasReady != nil && isReady != nil && isAnnotated {
-				log.Infof("pod update event skipped: added/labeled by CNI plugin")
+		}
+
+		// At this point, we know the pod
+		// - Looks like it should be enrolled
+		// - Looks like it isn't currently enrolled
+		//
+		// So we do a few checks to make sure the pod hasn't already been handled by the CNI plugin
+		// (which should always win over the informer), and actually has an IP.
+
+		// To things have to be true/asserted before the informer can enroll a pod:
+		// - The informer cannot assume the CNI plugin has handled a starting pod.
+		//   It could have been "stuck" in a starting state before the plugin was put in place.
+		//   Or, we could get unlucky with namespace labeling timing, and we could get an informer
+		//   update event for a pod immediately _after_ the plugin already checked that it was disabled
+		//   and skipped it.
+		//
+		//  - If the CNI plugin *has* handled a starting pod, and this is a stale event,
+		//  the informer must ignore it. This is confirmed by checking for annotation on the pod.
+		//
+		//  Note that this is not _completely_ raceproof in theory, but it is in practice.
+		//  Making adds idempotent (ipset, iptables, ztunnel) is required to obviate the
+		//  need for making adds raceproof in theory, and this could be removed when/if that idempotency
+		//  is in place.
+		if !kube.CheckPodRunning(newPod) {
+			alreadyHandled, err := s.GetPodIfAmbientActive(newPod.Name, newPod.Namespace)
```

Comment:
No, because this is doing `s.Pods.Get` and checking the annotation status of that, and `isAnnotated` is just checking the pod struct we got in the _event_ is annotated.